File: modules/text-recognition/android/src/main/java/expo/modules/textrecognition/TextRecognitionModule.kt
package expo.modules.textrecognition

import android.net.Uri
import android.content.Context
import android.graphics.Rect
import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.exception.CodedException
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import java.io.IOException

private class TextRecognitionFailedException :
  CodedException("Text Recognition is failed")

class TextRecognitionModule : Module() {  
  override fun definition() = ModuleDefinition {
    Name("TextRecognition")

    AsyncFunction("recognizeImage") { localUri: String, promise: Promise ->
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()

      val uri = Uri.parse(localUri) 
      val image: InputImage
      try {
        image = InputImage.fromFilePath(context, uri)
        val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
        recognizer.process(image)
        .addOnSuccessListener { result ->
          val response = mapOf(
            "width" to image.getWidth(),
            "height" to image.getHeight(),
            "blocks" to run {
              val blocks = mutableListOf<Any>()
              for (block in result.textBlocks) {
                val blockText = block.text
                val blockFrame = block.boundingBox
                val blockObject = mutableMapOf<String, Any>("blockText" to blockText, "blockFrame" to getRectMap(blockFrame))
                val lines = mutableListOf<Any>()
                for (line in block.lines) {
                  val lineText = line.text
                  val lineFrame = line.boundingBox
                  val lineObject = mapOf<String, Any>("lineText" to lineText, "lineFrame" to getRectMap(lineFrame))
                  lines.add(lineObject)
                }
                blockObject["lines"] = lines
                blocks.add(blockObject)
              }
              blocks
            }
          )
          promise.resolve(response)
        }
        .addOnFailureListener { e ->
          promise.reject("ERR_TEXT_TECOGNITION_EXCEPTION", "Text Recognition is failed.", e)
        }
      } catch (e: IOException) {
        e.printStackTrace()
        throw TextRecognitionFailedException()
      }
      Log.d("TextRecognitionModule", "Url: " + localUri); //to confirm that the file uri we passed from JS side is gotten here :)
    }
  }

  //Helper function to convert a Rect Type to a readable map
  private fun getRectMap(rect: Rect?): MutableMap<String, Int>{
    val rectObject = mutableMapOf<String, Int>()

     if (rect == null){
      return rectObject
    }

    rectObject["left"] = rect.left
    rectObject["top"] = rect.right
    rectObject["width"] = rect.right - rect.left //coordinates start from left to right
    rectObject["height"] = rect.bottom - rect.top //coordinates start from top to bottom

    return rectObject;
  }
}

File: modules/text-recognition/ios/TextRecognitionModule.swift
import ExpoModulesCore
import MLKitTextRecognition
import MLKitVision

public class TextRecognitionModule: Module {
  public func definition() -> ModuleDefinition {
    Name("TextRecognition")

    AsyncFunction("recognizeImage") { (localUri: String, promise: Promise) in
      let fileURL = URL(string:localUri)! // Let is defining a constant variable
      let imageData = try? Data(contentsOf: fileURL)
      if let mImage = UIImage(data: imageData!){
        let visionImage = VisionImage(image: mImage)
        visionImage.orientation = mImage.imageOrientation
        let latinOptions = TextRecognizerOptions()
        let latinTextRecognizer = TextRecognizer.textRecognizer(options: latinOptions)
        latinTextRecognizer.process(visionImage) { result, error in
          guard error == nil, let result = result else {
            // Error handling
            return promise.reject(TextRecognitionFailedException(error?.localizedDescription))
          }
          // Recognized text
          var response = ["width": mImage.size.width, "height": mImage.size.height, "blocks": []] 
          for block in result.blocks {
            let blockText = block.text
            let blockFrame = block.frame
            var blockObject = ["blockText": blockText, "blockFrame": self.getRectMap(rect: blockFrame)] 
            var lines = []
            for line in block.lines {
              let lineText = line.text
              let lineFrame = line.frame
              let lineObject = ["lineText": lineText, "lineFrame": self.getRectMap(rect: lineFrame)] 
              lines.append(lineObject)
            }
            blockObject["lines"] = lines
            var existingBlocks = response["blocks"] as? [[String: Any]] ?? [[String: Any]]()
            existingBlocks.append(blockObject)
            response["blocks"] = existingBlocks
          }
          return promise.resolve(response)
        }
      } else {
        return promise.reject(FileNotFoundException(localUri))
      }
    }
  }

  private func getRectMap(rect: CGRect?) -> Dictionary<String, CGFloat> {
    var rectObject: [String: CGFloat] = [:] //initialize to an emptyMap. The var makes it mutable.

    rectObject["left"] = rect?.origin.x
    rectObject["top"] = rect?.origin.y
    rectObject["width"] = rect?.size.width
    rectObject["height"] = rect?.size.height

    return rectObject
  }
}
